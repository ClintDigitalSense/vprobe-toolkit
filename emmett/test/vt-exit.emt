# ##########################################################
# Copyright 2007 VMware, Inc.  All rights reserved.
# -- VMware Confidential
# ##########################################################/

string str;

# Pretty-printer for VT exits
string 
vt_pretty_print() {
   code = VMCS_EXIT_REASON;
   sprintf(str, "%3d ", code);
   # Exit code 0: software exception
   if (code == 0) {
      intrInfo = VMCS_EXIT_INTR_INFO;
      faultNum = intrInfo & 0xff;
      sprintf(str, "%s softint exc %d", str, faultNum);
      if (faultNum == 14) {
         sprintf(str, "%s #PF @%#llx", str, VMCS_EXIT_QUAL);
      }
   }
   if (code == 1) {
      intrInfo = VMCS_EXIT_INTR_INFO;
      sprintf(str, "%s hwIRQ %03d", str, intrInfo & 0xff);
   }
   if (code == 12) {
      sprintf(str, "%s HLT", str);
   }
   if (code == 14) {
      sprintf(str, "%s INVLPG @%#llx", str, VMCS_EXIT_QUAL);
   }
   # Exit code 28: CR
   if (code == 28) {
      qual = VMCS_EXIT_QUAL;
      sprintf(str, "%s CR%1d", str, qual & 0xf);
      if (qual & (1 << 4)) {
         sprintf(str, "%s WRITE ", str);
      } else {
         sprintf(str, "%s READ  ", str);
      }
   }
   
   # Exit code 30: PIO instruction
   if (code == 30) {
      qual = VMCS_EXIT_QUAL;
      wasIn = qual & (1 << 3);
      port  = (qual >> 16) & 0xfff;
      if (wasIn) {
         sprintf(str, "%s %s", str, "IN  ");
      } else {
         sprintf(str, "%s %s", str, "OUT ");
      }
      sprintf(str, "%s port %04x", str, port);
   }
   if (code == 43) {
      sprintf(str, "%s TPR", str);
   }
   return str;
}

# E.g.:
# HV_Exit exits[vt_pretty_print()]++;
