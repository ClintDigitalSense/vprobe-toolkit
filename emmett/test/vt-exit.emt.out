(version 0.2)
(definteger code)
(definteger faultNum)
(definteger intrInfo)
(definteger port)
(definteger qual)
(defstring str)
(definteger wasIn)
(defun vt_pretty_print()( do  ( setint code   VMCS_EXIT_REASON)( sprintf    str  "%3d "  code )( cond  ( (== code  0 )( do  ( setint intrInfo   VMCS_EXIT_INTR_INFO)( setint faultNum    (& intrInfo  255 ))( sprintf    str  "%s softint exc %d"  str  faultNum )( cond  ( (== faultNum  14 )( do  ( sprintf    str  "%s #PF @%#llx"  str  VMCS_EXIT_QUAL )))))))( cond  ( (== code  1 )( do  ( setint intrInfo   VMCS_EXIT_INTR_INFO)( sprintf    str  "%s hwIRQ %03d"  str   (& intrInfo  255 ) ))))( cond  ( (== code  12 )( do  ( sprintf    str  "%s HLT"  str ))))( cond  ( (== code  14 )( do  ( sprintf    str  "%s INVLPG @%#llx"  str  VMCS_EXIT_QUAL ))))( cond  ( (== code  28 )( do  ( setint qual   VMCS_EXIT_QUAL)( sprintf    str  "%s CR%1d"  str   (& qual  15 ) )( cond  ( (& qual   (<< 1  4 ) )( do  ( sprintf    str  "%s WRITE "  str ))) (1 ( do  ( sprintf    str  "%s READ  "  str )))))))( cond  ( (== code  30 )( do  ( setint qual   VMCS_EXIT_QUAL)( setint wasIn    (& qual   (<< 1  3 ) ))( setint port    (&  (>> qual  16 )  4095 ))( cond  (wasIn( do  ( sprintf    str  "%s %s"  str  "IN  " ))) (1 ( do  ( sprintf    str  "%s %s"  str  "OUT " ))))( sprintf    str  "%s port %04x"  str  port ))))( cond  ( (== code  43 )( do  ( sprintf    str  "%s TPR"  str )))))str)